beacon/light/committee_chain.go

```diff
@@ -0,0 +1,587 @@
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	if !s.committees.periods.CanExpand(period) {
+		return ErrInvalidPeriod
+	}
+	root := s.getCommitteeRoot(period)
+	if root == (common.Hash{}) {
+		return ErrInvalidPeriod
+	}
+	if root != committee.Root() {
+		return ErrWrongCommitteeRoot
+	}
+	if !s.committees.periods.Includes(period) {
+		if err := s.committees.add(s.db, period, committee); err != nil {
+			return err
+		}
+		s.syncCommitteeCache.Remove(period)
+	}
+	return nil
+}
+
+// GetUpdate returns the update at the given period.
+// Note: GetUpdate can be called either with locked or unlocked chain mutex.
+func (s *CommitteeChain) GetUpdate(period uint64) *types.LightClientUpdate {
+	update, _ := s.updates.get(period)
+	return update
+}
+
+// InsertUpdate adds a new update if possible.
+func (s *CommitteeChain) InsertUpdate(update *types.LightClientUpdate, nextCommittee *types.SerializedSyncCommittee) error {
+	s.chainmu.Lock()
+	defer s.chainmu.Unlock()
+
+	period := update.AttestedHeader.Header.SyncPeriod()
+	if !s.updates.periods.CanExpand(period) || !s.committees.periods.Includes(period) {
+		return ErrInvalidPeriod
+	}
+	if s.minimumUpdateScore.BetterThan(update.Score()) {
+		return ErrInvalidUpdate
+	}
+	oldRoot := s.getCommitteeRoot(period + 1)
+	reorg := oldRoot != (common.Hash{}) && oldRoot != update.NextSyncCommitteeRoot
+	if oldUpdate, ok := s.updates.get(period); ok && !update.Score().BetterThan(oldUpdate.Score()) {
+		// a better or equal update already exists; no changes, only fail if new one tried to reorg
+		if reorg {
+			return ErrCannotReorg
+		}
+		return nil
+	}
+	if s.fixedRoots.periods.Includes(period+1) && reorg {
+		return ErrCannotReorg
+	}
+	if !s.verifyUpdate(update) {
+		return ErrInvalidUpdate
+	}
+	addCommittee := !s.committees.periods.Includes(period+1) || reorg
+	if addCommittee {
+		if nextCommittee == nil {
+			return ErrNeedCommittee
+		}
+		if nextCommittee.Root() != update.NextSyncCommitteeRoot {
+			return ErrWrongCommitteeRoot
+		}
+	}
+	batch := s.db.NewBatch()
+	if reorg {
+		s.rollback(batch, period+1)
+	}
+	if addCommittee {
+		if err := s.committees.add(batch, period+1, nextCommittee); err != nil {
+			return err
+		}
+		s.syncCommitteeCache.Remove(period + 1)
+	}
+	if err := s.updates.add(batch, period, update); err != nil {
+		return err
+	}
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	log.Info("Inserted new committee update", "period", period, "next committee root", update.NextSyncCommitteeRoot)
+	return nil
+}
+
+// NextSyncPeriod returns the next period where an update can be added and also
+// whether the chain is initialized at all.
+func (s *CommitteeChain) NextSyncPeriod() (uint64, bool) {
+	s.chainmu.RLock()
+	defer s.chainmu.RUnlock()
+
+	if s.committees.periods.IsEmpty() {
+		return 0, false
+	}
+	if !s.updates.periods.IsEmpty() {
```

Comment:
It's a bit redundant, for a valid committee chain, the period of last committee always equals to the next period of the update.

Reply:
Actually now it's not true, the chain allows adding committees with fixed roots at the end. This never happens though so we can just as well make the chain stricter.

Reply:
I tried to make the logic stricter so that if there are updates then we don't allow committees after `updates.Next`but I reverted it because I realized it would just make everything messier. When the chain is initialized there is a committee but no updates yet. Here it would save one check but at `AddCommittee`, `InsertUpdate` and `checkConstraints` (and also in the syncing logic, if there is backward syncing) it would require extra conditions.